Add support to use heap commands without debug symbols

lebr0nli opened the pull request:
I'm trying to make some features that #937 mentioned.

I use two classes to represent different heap resolving strategies(by debug symbols or heuristic), and use some classes in `pwndbg/heap/structs.py` to make some ctypes struct compatible with `gdb.Value` and `gdb.Type`.

We can use `set resolve-via-heuristic on/off` to enable/disable the feature in this PR.

Hope this PR is useful :)

----

P.S. ~Some features only work for i386 and x86-64 currently~, and their implements are a little bit dirty(but it seems to work). I'm sorry about that :(
Edit: It supports arm/aarch64 now

lebr0nli committed:
Add support to use heap commands without debug symbols

disconnect3d commented:
Hey, thanks a lot for working on this <3. I owe you a 🍺 :D

I will try to look into it sooner or later and maybe implement some small tests for x86, x64 and arm if possible 🤐  (if you have ideas how to test it or want to implement tests urself, then go ahead!)

CptGibbon commented:
This is some cracking work, very creative 🤩

The only improvement I can see is that the `HeuristicHeap` version of `get_heap_boundaries()` doesn't attempt to resolve the `mp_` stuct before checking if `self._mp_addr` is populated. This means when the [heap] vm region and the actual heap start address are different, commands like "vis_heap_chunks" and "heap" will fail. A user can work around this by resolving the `mp_` struct using the "mp" command before issuing further heap commands.

Also a small typo causing issues, there are 2 members of your `c_malloc_par_2_25` struct named `mmaped_mem` & `max_mmaped_mem` which I think should be spelled "mmapped", although please correct me if I'm wrong.

Many thanks for adding this great functionality 🙏

lebr0nli commented:
@disconnect3d 
Hi, sorry to be replying so late. I'm trying to implement the unit test for x86 and x64 now, but I didn't have experience in writing unit tests, so it might take me some time to learn it. So if you finish your implements first, please use them directly XD
By the way, if you have any ideas or suggestions, feel free to tell me, too!

@CptGibbon 
Hi, you are right! Thanks for pointing it out!
I'll fix them in the future commit!

lebr0nli committed:
Fix a possible bugs when getting heap boundaries via heuristic

See https://github.com/pwndbg/pwndbg/pull/1029#issuecomment-1189841299

lebr0nli committed:
Fix typo causing issues in `c_malloc_par_2_25`

See https://github.com/pwndbg/pwndbg/pull/1029#issuecomment-1189841299

CptGibbon commented:
Awesome! Those fixes work with my testing 👍

All I can see that might need attention is the addition of another `TcachePerThreadStruct` that supports the pre-2.30 `count` fields, which are 8 bytes wide on x86_64 rather than 16 bytes. Right now commands like "tcachebins" print incorrect values with GLIBC builds that use these narrower `count` fields.

Just one more tcache-related issue; the heuristic method assumes a thread's tcache resides at the start of a heap region, which although common may not be the case once arena sharing occurs. I'm not sure how you would resolve a thread's `tcache` symbol heuristically though, and frankly this whole PR seems like magic to me already 🪄 so I think it's a reasonable compromise.

Thanks again.

lebr0nli commented:
> pre-2.30 `count` fields

I didn't notice that GLIBC < 2.30 is using `char counts[TCACHE_MAX_BINS];` instead of `uint16_t counts[TCACHE_MAX_BINS];`, thank you for pointing it out again 🙏
I'll fix it.

> although common may not be the case once arena sharing occurs

Oh, I forgot about this possibility, I'll take a look at the source code of `malloc.c` and try to figure out if we can find its real location like what we do on `thread_arena` currently.
Thanks!

lebr0nli committed:
Fix a bug for `tcache_perthread_struct` and refactor some codes in `structs.py`

* The bug: `tcache_perthread_struct` for GLIBC < 2.30 is using `char` instead of `uint16_t` for `counts` field

lebr0nli committed:
Fix some bugs about handling `thread_arena` and `tcache` with multithreaded

disconnect3d commented:
Regarding testing, would something like qemu-user emulation be sufficient here? Then we could run e.g. arm32, aarch64, etc binaries and test different cases?

lebr0nli commented:
As I mentioned before, I'm not adding support to arm32 and aarch64 in my PR yet 😢, so I'm not sure about qemu-user is sufficient or not.
After I have enough knowledge about arm architecture, I'll try to add support to them if possible!(maybe need a few weeks to learn about it, not sure how long it will take)

disconnect3d commented on pwndbg/heap/structs.py:
```diff
@@ -0,0 +1,593 @@
+BINMAPSIZE = 4
+TCACHE_MAX_BINS = 64
+
+if pwndbg.arch.ptrsize == 4:
```
This is problematic, we need to invoke this when arch changes in GDB

(One can do sth like `attach <pid>` to attach to different architecture in GDB)

lebr0nli replied:
I try to fix it in https://github.com/pwndbg/pwndbg/pull/1029/commits/d89e484ba0e47b44fad470a5e29dbc9add2113d2
I re-initialize the heap when `pwndbg.event.stop`, and I add a decorator to reload `pwndbg.heap.structs` every time we create a new instance of `HeuristicHeap`.
Does this fix the problem?

disconnect3d commented:
Hmm do we have then some guard code against other architectures than x86/x64?

CptGibbon commented:
The tcache commits are working nicely, no more issues that I can see.
Thanks for this PR 🙏

lebr0nli committed:
Re-initialize the heap when the process stop or the file changed

By doing this, we can attach to another architecture in GDB without any bugs.

lebr0nli commented:
> Hmm do we have then some guard code against other architectures than x86/x64?

No, we don't have it.
I try to add them in https://github.com/pwndbg/pwndbg/pull/1029/commits/d924d988be9a0b85ea4f21431ac6e4d245f68a88 now.

lebr0nli committed:
Add guard code for unsupported architectures

lebr0nli committed:
Support heuristic for arm and aarch64

Note: thread_arena and thread_cache for arm still can not work

lebr0nli commented:
I add support to arm32 and aarch64 in https://github.com/pwndbg/pwndbg/pull/1029/commits/8b5d15e2f16b7058b1d71a4bcbe7d1324a2fa840 now.

But still have some problems:

* I didn't test it on many GLIBC versions because it's a little bit difficult for me to debugging with QEMU. But at least for the arm32/aarch64 version GLIBC 2.31 which you can get via `apt-get`, it can work. So if the other version of GLIBC's assembly code pattern didn't have too much different, it should work, too. This part might need someone who has a real arm32/aarch64 device or who is familiar with arm to test it.

* ~I didn't find a way to get tls base of arm32(which means we can not find arena and tcache when we are not in the main thread), because it access a coprocessor register. And I didn't find a way to access it in GDB currently.~
Edit: I found a way to get tls base for arm via using the code gadget inside `__errno_location` now, but I think it still needs a better way to find tls for arm though. See https://github.com/pwndbg/pwndbg/pull/1029/commits/19b42c33aba3303fd9ea3cbd36b2ceceef45f768 and https://github.com/pwndbg/pwndbg/pull/1029/commits/8f6dc355faac35b6b924c70735d0932eddb20236

And after some testing, the answer of the question:

> Regarding testing, would something like qemu-user emulation be sufficient here? Then we could run e.g. arm32, aarch64, etc binaries and test different cases?

Yes, I guess qemu-user is sufficient to test it, but the vmmap need to work for QEMU.
To anyone want to test this PR with QEMU, you can re-compile @disconnect3d ’s patch for QEMU via:

```shell
git clone --branch patchew/20220221030910.3203063-1-dominik.b.czarnota@gmail.com --single-branch https://github.com/patchew-project/qemu.git
cd qemu
git submodule update --init --recursive
./configure --gdb=/usr/local/bin/gdb --enable-debug
make -j$(nproc) && make install
```

And then just remove some `is qemu` check in some `vmmap` related function. By doing this, vmmap won't work very perfectly, but it's enough to make it work.

lebr0nli committed:
Update .pylintrc

* Ignore `import-error` error for `import gdb`
* Ignore `no-member` error for `pwndbg.typeinfo.*`, because most of its members are dynamically generated.
* Ignore `protected-access` warning for `_fields_`, `_type_`, `_length_`, because ctypes don't have other ways to access them.

lebr0nli committed:
Refactor some code and comment to fit pep8 and lint check

lebr0nli committed:
Add a feature to enable users set symbol addresses manually

For example, by using `set main_arena 0xdeadbeaf`, pwndbg will try to find main_arena at 0xdeadbeaf when using heuristic

lebr0nli committed:
Use `__errno_location` to find TLS base for arm

By doing this, we can get `thread_arena` and `tcache` address

lebr0nli committed:
Block other thread before `__errno_location()`

lebr0nli committed:
Fix a bug for arm32 and a typo-caused bug

* Some wrong field names inside `c_heap_info` may cause some bugs in the future if we want to access it

* `pad` size of `heap_info` for arm32 is 0 byte, only i386 is 8 bytes, so I fixed it in a hard-coded way temporary

lebr0nli committed:
Fix #1044 related issues

disconnect3d commented on pwndbg/commands/__init__.py:
```diff
@@ -259,7 +259,7 @@ def _OnlyAmd64(*a, **kw):
 def OnlyWithLibcDebugSyms(function):
```
We should rename it to `OnlyWithResolvedHeapSyms` or `OnlyWithHeap` or some other name if u have a better one

lebr0nli replied:
I agree with yours.
I think `OnlyWithResolvedHeapSyms` is good, I'll use it.

disconnect3d commented on pwndbg/commands/__init__.py:
```diff
@@ -259,7 +259,7 @@ def _OnlyAmd64(*a, **kw):
+        if pwndbg.heap.current.libc_has_debug_syms() or pwndbg.config.resolve_via_heuristic:
             return function(*a, **kw)
         else:
             print('''%s: This command only works with libc debug symbols.
```
We should improve the messaging here, but we can work it out on Discord after merging this

lebr0nli replied:
Sure!
You mean the hint messaging for it, right?

disconnect3d replied:
yeah

disconnect3d commented on pwndbg/heap/__init__.py:
```diff
@@ -1,12 +1,81 @@
 heap_chain_limit = pwndbg.config.Parameter('heap-dereference-limit', 8, 'number of bins to dereference')
 
+resolve_via_heuristic = pwndbg.config.Parameter('resolve-via-heuristic', False,
+                                                'Resolve some missing symbols via heuristics')
```
```suggestion
                                                'Resolve missing heap related symbols via heuristics')
```

lebr0nli replied:
Sure, I'll try to fix it!

disconnect3d commented on pwndbg/heap/__init__.py:
```diff
@@ -1,12 +1,81 @@
 
 heap_chain_limit = pwndbg.config.Parameter('heap-dereference-limit', 8, 'number of bins to dereference')
 
+resolve_via_heuristic = pwndbg.config.Parameter('resolve-via-heuristic', False,
```
```suggestion
resolve_via_heuristic = pwndbg.config.Parameter('resolve-heap-via-heuristic', False,
```

lebr0nli replied:
Sure, I'll try to fix it!

disconnect3d commented on pwndbg/heap/__init__.py:
```diff
@@ -1,12 +1,81 @@
+resolve_via_heuristic = pwndbg.config.Parameter('resolve-via-heuristic', False,
+                                                'Resolve some missing symbols via heuristics')
+
+main_arena = pwndbg.config.Parameter('main_arena', "0", 'main_arena address for heuristics')
```
Hmm... those kinda bloats the config parameters a bit. Maybe we should do sth like "heapconfig" namespace and put all heap related params there?

I am fine with leaving them in config for now if u don't want to deal with it, we just have to make it clear what they mean/are for in their description --- I'd add text that they are related to heap or sth

lebr0nli replied:
Sure, I'll try to fix it! (with scope)

lebr0nli replied:
I added a new command called `heap_config` to show configs in `heap` scope(https://github.com/pwndbg/pwndbg/pull/1029/commits/289010da1dd4839bdb628b1f83390bbc352f88d5), does this fix the issue?

disconnect3d commented on pwndbg/heap/__init__.py:
```diff
@@ -1,12 +1,81 @@
+            elif address_str.startswith("0o"):
+                address = int(address_str, 8)
+            elif address_str.startswith("0x"):
+                address = int(address_str, 16)
```
All those if/elif here can be replaced with `int(address_str, 0)` - it will determine the format

lebr0nli replied:
Wow, I didn't know this cool feature before!
Thanks, I'll fix it!

disconnect3d commented on pwndbg/heap/__init__.py:
```diff
@@ -1,12 +1,81 @@
+
+@pwndbg.events.stop
+@pwndbg.events.new_objfile
+def reset():
```
What's the reason to reset on each stop?

lebr0nli replied:
That's because I want to make sure we refresh the heap like we update `pwndbg.typeinfo` when the file changes or process stopped, so we can avoid using some old addresses or wrong structs definitions.(btw my commit about it is https://github.com/pwndbg/pwndbg/commit/d89e484ba0e47b44fad470a5e29dbc9add2113d2)
But a little side effect is that not only when the file we want to debug changed, when loading some shared object files, the heap will keep refreshing too, and this wastes some resources.
So I think this part might need a better implementation, do you have any suggestions about this, or should I keep this part?
BTW, the way I use `importlib.reload` to reload structs seems a sort of dirty to me, should I write all the variables in an `update()` like what we did in `pwndbg.typeinfo`? It seems a little bit cleaner to me, but I'm not sure:
https://github.com/pwndbg/pwndbg/blob/e867d1db0c51017d5df927d86c33dd6b9b733456/pwndbg/typeinfo.py#L42-L47

disconnect3d replied:
RE: better idea for this - not on top of my head, I will try to see if we have anything like that elsewhere.
\
Another idea could be only to mark that heap is "dirty" here and refresh it only on demand when someone asks for information regarding it. I guess that would be better?

disconnect3d replied:
Regarding `importlib.reload` - yeah, I also think it is not the best approach, but we also do it in one of `update`, to reset elf types or something like that, if I recall correctly.

(I'm totally up for not using `reload` at all in those cases, so if u want to move it to `update` I agree on that)

lebr0nli replied:
> refresh it only on demand when someone asks for information regarding it. I guess that would be better?

Yes, I agree!
But sadly, after a few days of testing, I didn't figure out a good way to do it 😢
I'll update this part if I have a better ideas in the future.

> Regarding `importlib.reload` - yeah, I also think it is not the best approach, but we also do it in one of `update`, to reset elf types or something like that, if I recall correctly.
> 
> (I'm totally up for not using `reload` at all in those cases, so if u want to move it to `update` I agree on that)

Ok, I see.
And after some testing, I found that using `update()` might cause some errors in this case actually, so I'll keep the part of `importlib.reload`!

lebr0nli committed:
Refactor the code about heap related config

* Use `int(address_str, 0)` to auto determine the format (https://github.com/pwndbg/pwndbg/pull/1029#discussion_r939569382)

* Use `OnlyWithResolvedHeapSyms` instead of `OnlyWithLibcDebugSyms` (https://github.com/pwndbg/pwndbg/pull/1029#discussion_r939568687)

* Use `resolve-heap-via-heuristic` instead of `resolve-via-heuristic` (https://github.com/pwndbg/pwndbg/pull/1029#discussion_r939569076)

* Update the description of `resolve-heap-via-heuristic` config (https://github.com/pwndbg/pwndbg/pull/1029#discussion_r939569069)

* Move heap related config into `heap` scope, and add a new command, `heap_config`, to show the config in that scope (https://github.com/pwndbg/pwndbg/pull/1029#discussion_r939569260)

lebr0nli committed:
Refactor code about the config of heap related symbols

lebr0nli committed:
Fix the logic when thread_arena is not found

lebr0nli committed:
Use errno trick as a fallback for i386 and x86-64

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+            return self.main_arena
+
+        try:
+            next(i for i in pwndbg.vmmap.get() if arena_addr in i)
```
@lebr0nli What's the point of this `next(...)`?

lebr0nli replied:
It want to make sure the arena address is a valid address in memory (when `next` failed, it means we can't find it in memory)?
But it seems a little bit expensive because it use exception to handle it, maybe I should use `pwndbg.vmmap.find(arena_addr) is None` to determine it?
I'm not sure, I just refactor some of the original methods from `Heap` to `DebugSymsHeap` 😆

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+            return self.malloc_state(main_arena_via_config)
+        # TODO/FIXME: These are quite dirty, we should find a better way to do this
+        if not self._main_arena_addr:
+            if pwndbg.glibc.get_version() < (2, 34) and pwndbg.arch.current != "arm":
```
Just to be sure: what happens for "arm"? Is arm 32-bit fine?

lebr0nli replied:
Hmm, GEF's heuristic for `main_arena` seems to work fine for i386, x86-64, aarch64 with GLIBC < 2.34, but I'm not sure what happened on the heuristic of GEF for 32-bit arm actually, so I decided to use our version of heuristic.
And just guessing, maybe it's because GEF got the wrong `malloc_state` size for arm 32-bit?
You might reproduce it with a normal GLIBC 2.31 which is installed via `apt-get install arm-linux-gnueabihf-gcc`.

Here is a example, if we try the code from https://github.com/hugsy/gef/blob/dev/gef.py#L1180

```
gef➤  pi
>>> ctypes.sizeof(GlibcArena.malloc_state_t())
1116
>>> malloc_hook_addr = parse_address("(void *)&__malloc_hook")
>>> struct_size = ctypes.sizeof(GlibcArena.malloc_state_t())
>>> hex(malloc_hook_addr - gef.arch.ptrsize - struct_size)
'0x3f7bf76c'
>>>
gef➤  ptype/o main_arena
/* offset      |    size */  type = struct malloc_state {
/*      0      |       4 */    __libc_lock_t mutex;
/*      4      |       4 */    int flags;
/*      8      |       4 */    int have_fastchunks;
/*     12      |      40 */    mfastbinptr fastbinsY[10];
/*     52      |       4 */    mchunkptr top;
/*     56      |       4 */    mchunkptr last_remainder;
/*     60      |    1016 */    mchunkptr bins[254];
/*   1076      |      16 */    unsigned int binmap[4];
/*   1092      |       4 */    struct malloc_state *next;
/*   1096      |       4 */    struct malloc_state *next_free;
/*   1100      |       4 */    size_t attached_threads;
/*   1104      |       4 */    size_t system_mem;
/*   1108      |       4 */    size_t max_system_mem;

                               /* total size (bytes): 1112 */
                             }
gef➤  p &main_arena
$1 = (struct malloc_state *) 0x3f7bf76c <main_arena>
```
As you can see, idk why, GEF's heuristic got a correct `main_arena` address with a wrong size of `malloc_state` lol
But because our size is correct, it will failed.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                                self._main_arena_addr = offset + instr.address + 4
+                else:
+                    inform_report_issue("main_arena")
+                    raise OSError("Cannot find the symbol via heuristics")
```
Does this make sense? We have the last resort check of finding `main_arena` via _IO_* symbols in libc in lines 742+ --- shall we really raise an exception here instead of trying out a different approach?

disconnect3d replied:
We have an "else" with the same code at the end of the whole function, so maybe we should not fallback here and try more methods

lebr0nli replied:
Make sense!
Thanks, I'll fix it.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                if not self._main_arena_addr:
+                    # there are more than one arena, try to find by main_arena.top and main_arena.max_system_mem
+                    heap_page = next(x for x in pwndbg.vmmap.get() if "heap]" in x.objfile)
+                    for addr in range(_IO_2_1_stdin_addr, _IO_list_all_addr, pwndbg.arch.ptrsize):
```
How does `main_arena->next` work? I mean, I can see that when there is only one arena, the `main_arena->next == main_arena`. If we have two arenas, is it cyclic like `main_arena->next->next == main_arena`? If so, why don't we try to resolve it like this?

lebr0nli replied:
Good idea!
I didn't think about this before, I'll try to support what you said.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+        return self.is_initialized() and pwndbg.glibc.get_version() >= (2, 26)
+
+    def get_tls_base_via_errno_location(self):
+        assert pwndbg.arch.current in ("x86-64", "i386", "arm")
```
@lebr0nli can we add here some comment about Aarch64, as in why we don't have it here?

From https://maskray.me/blog/2021-02-14-all-about-thread-local-storage it seems TLS exists on aarch64... (well, why would it not :D)

lebr0nli replied:
Yeah, aarch64 has TLS, too.
But the `TPIDR_EL0` register which is used for aarch64 TLS seems always work.
So I didn’t implement that for it.
Is it possible that `TPIDR_EL0` failed to lead us to find TLS base?
If the answer is yes, maybe we need to implement for aarch64, too.

For other architectures:
I use this function for arm32 because I can’t find a way to access the coprocessor register(if I didn't misunderstand it).
And I use this for x86/x64 because sometimes we can’t always find gsbase/fsbase since its implementation is limited. 

And sure, I'll add some comment about it.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+
+    def get_tls_base_via_errno_location(self):
+        assert pwndbg.arch.current in ("x86-64", "i386", "arm")
+        # TODO/FIXME: Need a better way to find tls
```
Perhaps we should move TLS finding to an external function too

lebr0nli replied:
Ok, should we move this part to a new file, for example, `pwndbg/tls.py` or something?
And maybe we can add a new command in `pwndbg/commands/tls.py`, which is called `tls` or `tls_base` to let user to use it?

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+        already_lock = gdb.parameter("scheduler-locking") == "on"
+        if not already_lock:
+            gdb.execute("set scheduler-locking on")
+        errno_addr = int(gdb.parse_and_eval("(int *)__errno_location()"))
```
Is there a reason we make the call and not fetch it by the `errno` symbol address?

disconnect3d replied:
Maybe we don't have errno symbol address sometimes lol

lebr0nli replied:
> Maybe we don't have errno symbol address sometimes lol

Yes, `errno` symbol doesn't seem to work for qemu-user and some weird cases I guess.
And I'm not sure why, the result of `pwndbg.symbol.address('errno')` will not the same as `(int *)__errno_location` sometimes.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+            elif pwndbg.arch.current == "i386":
+                for instr in __errno_location_instr:
+                    # Find something like: mov eax, dword ptr [eax + disp]
+                    # (disp is a negative value, and eax is come from one of the page address of libc)
```
```suggestion
                    # (disp is a negative value, and eax comes from one of the page address of libc)
```

disconnect3d committed:
Update pwndbg/heap/ptmalloc.py

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                    elif instr.mnemonic == 'ldr' and "[pc," in instr.op_str:
+                        regs[instr.operands[0].str] = instr
+            else:
+                inform_report_issue("mp_")
```
Do we need this else here if there is another one below?

lebr0nli replied:
Maybe don't, so more architectures might be supported by heuristic.
Thanks! I'll fix it.

disconnect3d commented:
@lebr0nli Merged, but please look at the comments above :) Thanks again!

lebr0nli commented:
No problem, I'll take a look!
But I'm a little bit busy this week, I'll reply to those at the night of my time zone(here are 6 am on Monday now haha) or this Wednesday when I have time!

lonnywong commented:
Thanks for this great feature.

Can we resolve heap via heuristic only without libc debug symbols? I want to set resolve-heap-via-heuristic on by default. If there is libc debug symbols, use it first. It would be better if there is a hint shows that current has no libc debug symbols and using heuristic, at the first heap command.

Now I have to manually choose heuristic or not.

https://github.com/pwndbg/pwndbg/blob/ec8addd60243d78e6e02202cc00c3d2ace330392/pwndbg/commands/heap.py#L30-L33

https://github.com/pwndbg/pwndbg/blob/ec8addd60243d78e6e02202cc00c3d2ace330392/pwndbg/heap/__init__.py#L54-L70

lebr0nli commented:
> Can we resolve heap via heuristic only without libc debug symbols? I want to set resolve-heap-via-heuristic on by default. If there is libc debug symbols, use it first.

Good suggestion, but instead of turning it on by default, maybe it will be better that we set it to `auto` by default, then `auto` will do what you said(use heuristic when there are no debug symbols), but we can still let the user turn it on or off manually?

@disconnect3d, what do you think about my comment on this and @lonnywong's suggestion?

disconnect3d commented:
@lebr0nli I wanted to propose what @lonnywong wrote: to use it by default when debug symbols are missing. But tbh I like your solution with `auto` more.

On the other hand, will users actually use it? Or is this mainly to let users to disable the heap heuristics resolution?

CptGibbon commented:
+1 for `auto`

lebr0nli commented:
@disconnect3d @CptGibbon Ok, I see!

> On the other hand, will users actually use it? Or is this mainly to let users to disable the heap heuristics resolution?

@disconnect3d, I think yes, but only used for some weird CTF challenges probably xD.

I mean, let's say we have a libc that has every debug symbol except `main_arena` and `struct malloc_state`, then we'll need to turn on heap heuristic explicitly to make heap commands work if we "only" turn on heap heuristic when there's no debug symbol. So that's why I prefer `auto` instead of `on` more.
(But yeah, this scenario is a little bit over design I guess lol)

lonnywong commented:
How about `off`, `auto` and `force`? ( maybe over designed )

If there are no libc debug symbols for `auto` mode, I'd like to get a hint, maybe I'll try to install it.

lebr0nli commented:
> How about `off`, `auto` and `force`?

Oh, good idea, I just noticed that there's a parameter class called `gdb.PARAM_ENUM`, I'll take a look at whether `gdb.PARAM_ENUM` can help or not for this. (I only knew `gdb.PARAM_AUTO_BOOLEAN` before, so what I thought at first was that we can only use `on`/`off`/`auto` :P)

Btw I think `pwndbg.config.Parameter` might need some updates first because it only accepts a few classes of parameters, and we can't control the full content of `help set SOME_CONFIG_OF_PWNDBG` by customizing its `__doc__`.
I'll take a look at this in my free time, too.

disconnect3d commented:
> Btw I think pwndbg.config.Parameter might need some updates first because it only accepts a few classes of parameters, and we can't control the full content of help set SOME_CONFIG_OF_PWNDBG by customizing its __doc__.

:+1: we discussed somewhere else recently that I'd be great to enhance the config.Parameter class to support more robust help strings for parameters.

lebr0nli commented:
> 👍 we discussed somewhere else recently that I'd be great to enhance the config.Parameter class to support more robust help strings for parameters.

Oh, I saw the discussion about it in another PR now, that's great!

kunshim commented:
How can I use this feature? Is there special command need to use this feature?

lebr0nli commented:
> How can I use this feature? Is there special command need to use this feature?

In the latest pwndbg, by default, the heap heuristic(the feature you mentioned) should automatically be the fallback when pwndbg fails to use debug symbols to resolve the heap, so you don't need any special command to use/enable it.
But in some cases, you might need to manually set something more to make the heap heuristic work correctly.
For this, you can use `help set resolve-heap-via-heuristic` for more information about this feature.
